File: lib/approck-example/src/web/example/websocket1.ts
// create a websocket to /example/websocket1/ws
const ws = new WebSocket(window.location.href.replace(/^https:/, "wss:"));
const $output = document.getElementById("output") as HTMLDivElement;
const $send_foo_bar = document.getElementById("send-foo-bar") as HTMLButtonElement;

// print any messages
ws.onmessage = function(event) {
    console.log(event.data);
    let p = document.createElement("p");
    p.textContent = event.data;
    $output.appendChild(p);
};

ws.onerror = function(event) {
    console.error(event);
};

// send a message when the button is clicked
$send_foo_bar.onclick = function() {
    ws.send("foo bar");
};

File: lib/approck-example/src/web/example/websocket1.rs
#[approck::http(GET /example/websocket1; return HTML|WebSocketUpgrade;)]
pub mod page {
    pub async fn request(req: Request, ui: Document) -> Response {
        // An example of moving a variable into a closure which is used to produce the
        // websocket_handler future.
        let name = "jason";
        match req
            .upgrade_to_websocket(move |socket| async move {
                println!("{name}");
                self::websocket(socket).await
            })
            .await
        {
            Ok(Some(response)) => return Response::WebSocketUpgrade(response),
            Err(_error) => todo!(),
            _ => {}
        }

        ui.add_js("./websocket1.js");

        ui.add_body(maud::html! {
            div.container.bg-white {
                a href="/example/" { "◂ Back to Example List" }
                h1 { "WebSocket Example" }
                hr;
                p { "This example demonstrates a simple WebSocket connection." }
                p { "Open the browser console to see the messages." }
                hr;
                button #send-foo-bar { "Send FooBar" }
                hr;
                div #output {}
            }
        });

        Response::HTML(ui.into())
    }

    async fn websocket(mut websocket: WebSocket) {
        websocket
            .send("Hello from the server!".into())
            .await
            .unwrap();

        while let Some(message) = websocket.recv().await {
            let message = match message {
                Ok(message) => message,
                Err(error) => {
                    eprintln!("Error: {error:?}");
                    continue;
                }
            };
            match message.into_data() {
                WebSocketMessageData::Text(text) => {
                    println!("Received: {}", text);
                    websocket
                        .send(format!("Got this on server: {:?}", text).into())
                        .await
                        .unwrap();
                }
                WebSocketMessageData::Close => {
                    println!("Connection closed");
                }
                _ => {}
            }
        }
    }
}
